Implementation of grid uniform_channel_with_sphere

a412h opened the pull request:
Following the discussion at #18952, here is the implementation for the grid of a uniform channel that contains a sphere.
A typical set of parameters could be for example when used with a solver like ryujin:
```
subsection C - Discretization
    set geometry        = sphere_channel
    set mesh refinement = 5
    
    subsection sphere_channel  # In unit of cube length (2)
        set length before sphere = 1
        set length after sphere =  3
        set height below sphere =  1
        set height above sphere =  1
        set depth =                1
    end
end
```

blaisb commented on source/grid/grid_generator.cc:
```diff
@@ -9685,6 +9685,139 @@ namespace GridGenerator
+    const bool                       colorize)
+  {
+    const double radius = 0.5;
+    const double box_radius = 1;
```
The depth default value is equal to one. Are there limitations on the depth value with respect to the other parameter? Is the depth the number of subdivision in the depth or the physical depth in terms of number of sphere radius? :)

a412h commented:
Before I send a commit that addresses all points above, could you please use the following test and see if the function does what is expected ? To keep the mesh conformal in a simple way, the refinement is controlled only by deal function refine_global, as such there is no subdivision parameter. For the depth parameter, the same value is applied before and after the sphere in the z direction, but I can change that to create a sphere that is not centered in the z axis. The parameters to position the sphere are in units of the cube length (1 unit = 2 * box_radius):

[main.cpp](https://github.com/user-attachments/files/23730779/main.cpp)
[CMakeLists.txt](https://github.com/user-attachments/files/23730832/CMakeLists.txt)

a412h commented on include/deal.II/grid/grid_generator.h:
```diff
@@ -644,6 +644,18 @@ namespace GridGenerator
+  /**
+   * Generate a 3D grid consisting of a channel with a sphere where the length,
+   * the height and the depth of the channel can be defined by the user.
+   */
```
I updated the description following the way it is done for the case `uniform_channel_with_cylinder`.

blaisb replied:
Nicely done. thanks!

a412h commented on source/grid/grid_generator.cc:
```diff
@@ -9685,6 +9685,139 @@ namespace GridGenerator
+    tria.set_manifold(1, SphericalManifold<3>());
+    TransfiniteInterpolationManifold<3> transfinite_manifold;
+    transfinite_manifold.initialize(tria);
+    tria.set_manifold(0, transfinite_manifold);
```
I have made the changes; now there is:

1. A spherical manifold for the sphere,
2. If `use_transfinite_region` is `true`, the rest of the domain uses a `TransfiniteInterpolationManifold`,
3. Else the rest of the domain uses a `FlatManifold`.

blaisb commented on tests/grid/grid_generator_uniform_channel_with_sphere.cc:
```diff
@@ -0,0 +1,45 @@
+// ------------------------------------------------------------------------
+//
+// SPDX-License-Identifier: LGPL-2.1-or-later
+// Copyright (C) 2013 - 2018 by the deal.II authors
```
```suggestion
// Copyright (C) 2025 by the deal.II authors
```

blaisb reviewed (commented):
The only thing missing is the image of your grid and this will be ready to merge :)! Awesome and thank you for your patience!

a412h commented:
@blaisb I think it is ready for review. I have tried to make the new version more consistent with the function `uniform_channel_with_cylinder` (use of `hyper_shell`, then `concentric_hyper_shell`).

blaisb commented on include/deal.II/grid/grid_generator.h:
```diff
@@ -644,6 +644,110 @@ namespace GridGenerator
+   * This generator can be used for benchmarking Navier-Stokes solvers
+   * for various flows around a sphere cases in 3D.
+   * The sphere is centered at $(0, 0, 0)$, has diameter twice inner_radius and lies at the center of a cube
+   * of length twice outer_radius.
```
```suggestion
   * of length which is two times the outer_radius.
```

a412h commented:
@blaisb Thanks for the corrections, modifications pushed.
